File: agenticx/runtime/assistant_output.py
"""Canonical assistant output parser for think/followups protocol.

Single-pass tag state machine shared by Runtime, Studio, and Desktop mirrors.
Author: Damon Li
"""

from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from typing import Optional, Sequence, Tuple

# MiniMax models can leak mangled chat-template control tokens (e.g.
# ``]<]minimax[>[`` or ``<minimax_end>``) into generated text — including the
# tail of a <followups> line. Strip only the bracket/angle-wrapped forms so a
# legitimate plain "MiniMax" word in prose is never touched.
_MINIMAX_ARTIFACT_RE = re.compile(
    r"[\]\[<>~!|]+\s*/?\s*minimax[a-z_:]*\s*[\]\[<>~!|]*",
    re.IGNORECASE,
)

_ZERO_WIDTH_RE = re.compile(
    r"[\u200b\u200c\u200d\u2060\ufeff\u00ad]",
)
_RESERVED_TAG_TOKEN_RE = re.compile(
    r"</?\s*think\b|</?\s*followups\b",
    re.IGNORECASE,
)
_CONTROL_CHAR_RE = re.compile(
    r"[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]",
)

_TAG_NAME_RE = re.compile(
    r"^(?P<closing>/?)\s*"
    r"(?P<name>think|followups)\b"
    r"(?P<rest>.*)$",
    re.IGNORECASE | re.DOTALL,
)

_PROTOCOL_ERROR_ORDER = (
    "nested_reserved_tag",
    "mismatched_close",
    "stray_close",
    "unclosed_think",
    "unclosed_followups",
    "multiple_followups",
    "noncanonical_reserved_tag",
    "missing_visible_body",
    "invalid_followup_count",
    "invalid_followup_line",
)


@dataclass(frozen=True)
class ParsedAssistantOutput:
    """Normalized assistant turn: reasoning, body, suggestions, protocol errors."""

    reasoning: str
    visible_body: str
    suggested_questions: tuple[str, ...]
    protocol_errors: tuple[str, ...]

    @property
    def malformed(self) -> bool:
        return bool(self.protocol_errors)


def strip_model_control_artifacts(text: str) -> str:
    """Remove bracket-wrapped MiniMax control-token residue from *text*."""
    if not text or "minimax" not in text.lower():
        return text
    return _MINIMAX_ARTIFACT_RE.sub("", text).strip()


def sanitize_public_tool_summary(raw: str) -> str | None:
    """Return a user-safe public tool message, or None when unsafe."""
    if not isinstance(raw, str):
        return None
    text = raw.strip()
    if not text:
        return None
    if _ZERO_WIDTH_RE.search(text) or _CONTROL_CHAR_RE.search(text):
        return None
    if _RESERVED_TAG_TOKEN_RE.search(text):
        return None
    # Reject noncanonical reserved-tag shapes such as ``< Think >``.
    if _contains_noncanonical_reserved_tag(text):
        return None
    cleaned = strip_model_control_artifacts(text)
    if not cleaned or cleaned != text and _RESERVED_TAG_TOKEN_RE.search(cleaned):
        return None
    if not cleaned.strip():
        return None
    return cleaned


def _contains_noncanonical_reserved_tag(text: str) -> bool:
    idx = 0
    while True:
        start = text.find("<", idx)
        if start < 0:
            return False
        end = text.find(">", start + 1)
        if end < 0:
            return False
        inner = text[start + 1 : end]
        parsed = _classify_tag(inner)
        if parsed is not None and parsed[2]:
            return True
        idx = start + 1


def _is_markdown_artifact_line(line: str) -> bool:
    s = line.strip()
    if not s:
        return True
    if s.startswith("```") or s.endswith("```"):
        return True
    if s.startswith("#"):
        return True
    if re.match(r"^\|.*\|$", s):
        return True
    if re.match(r"^\|?\s*:?-{3,}:?\s*(\|\s*:?-{3,}:?\s*)+\|?$", s):
        return True
    return False


def _line_has_reserved_tag_token(line: str) -> bool:
    return bool(_RESERVED_TAG_TOKEN_RE.search(line))


def sanitize_suggested_questions(
    raw: object,
    visible_body: str,
) -> tuple[str, ...]:
    """Validate follow-up chip lines; return exactly three safe lines or empty."""
    if not str(visible_body or "").strip():
        return ()

    candidates: list[str]
    if isinstance(raw, str):
        candidates = raw.splitlines()
    elif isinstance(raw, Sequence) and not isinstance(raw, (bytes, bytearray)):
        candidates = [str(item) for item in raw]
    else:
        return ()

    lines: list[str] = []
    for item in candidates:
        trimmed = strip_model_control_artifacts(str(item).strip())
        if trimmed:
            lines.append(trimmed)

    if len(lines) != 3:
        return ()

    for line in lines:
        if not (1 <= len(line) <= 160):
            return ()
        if _line_has_reserved_tag_token(line):
            return ()
        if _is_markdown_artifact_line(line):
            return ()
    return tuple(lines)


def _normalize_tag_inner(inner: str) -> str:
    # Strip zero-width and collapse whitespace for tag-name recognition.
    cleaned = _ZERO_WIDTH_RE.sub("", inner)
    cleaned = "".join(
        ch for ch in cleaned if unicodedata.category(ch) != "Cf"
    )
    return cleaned.strip()


def _canonical_tag_text(name: str, closing: bool) -> str:
    body = f"/{name}" if closing else name
    return f"<{body}>"


def _classify_tag(inner: str) -> Optional[Tuple[str, bool, bool]]:
    """Return (name, is_closing, noncanonical) or None if not a reserved tag."""
    normalized = _normalize_tag_inner(inner)
    if not normalized:
        return None
    match = _TAG_NAME_RE.match(normalized)
    if not match:
        return None
    name = match.group("name").lower()
    closing = bool(match.group("closing"))
    rest = match.group("rest") or ""
    # Canonical form is exactly ``think`` / ``followups`` with optional trailing
    # spaces only (no attributes). Any extra spacing inside the original ``<>``
    # or attributes marks the tag noncanonical.
    compact_canonical = f"{'/' if closing else ''}{name}"
    original_compact = re.sub(r"\s+", "", _ZERO_WIDTH_RE.sub("", inner)).lower()
    noncanonical = original_compact != compact_canonical or bool(rest.strip())
    # Also treat leading/trailing spaces inside <> as noncanonical.
    if inner != inner.strip() or "\n" in inner or "\r" in inner:
        noncanonical = True
    if _ZERO_WIDTH_RE.search(inner):
        noncanonical = True
    return name, closing, noncanonical


class _ProtocolErrorSet:
    __slots__ = ("_items", "_seen")

    def __init__(self) -> None:
        self._items: list[str] = []
        self._seen: set[str] = set()

    def add(self, code: str) -> None:
        if code not in self._seen and code in _PROTOCOL_ERROR_ORDER:
            self._seen.add(code)
            self._items.append(code)

    def as_tuple(self) -> tuple[str, ...]:
        return tuple(self._items)


class AssistantOutputStreamParser:
    """Incremental tokenizer mirroring :func:`parse_assistant_output`."""

    __slots__ = (
        "_carry",
        "_stack",
        "_body",
        "_reasoning",
        "_followup_buf",
        "_followups_blocks",
        "_followups_invalid",
        "_discard_until",
        "_errors",
        "_safe_stream",
        "_emitted_len",
        "_think_open_emitted",
    )

    def __init__(self) -> None:
        self._carry = ""
        self._stack: list[str] = []
        self._body = ""
        self._reasoning = ""
        self._followup_buf = ""
        self._followups_blocks = 0
        self._followups_invalid = False
        self._discard_until: Optional[str] = None
        self._errors = _ProtocolErrorSet()
        self._safe_stream = ""
        self._emitted_len = 0
        self._think_open_emitted = False

    def feed(self, chunk: str) -> str:
        """Return only the newly safe-to-stream visible/reasoning delta."""
        if not chunk:
            return ""
        self._consume(self._carry + chunk, finalize=False)
        delta = self._safe_stream[self._emitted_len :]
        self._emitted_len = len(self._safe_stream)
        return delta

    def finalize(self) -> ParsedAssistantOutput:
        """Finalize the accumulated turn with the same tokenizer state."""
        if self._carry:
            # Unresolved partial ``<...`` prefix: treat as literal body text.
            self._emit_text(self._carry)
            self._carry = ""
        self._finalize_errors_and_suggestions()
        return self._to_parsed()

    def _consume(self, data: str, *, finalize: bool) -> None:
        i = 0
        n = len(data)
        self._carry = ""
        while i < n:
            ch = data[i]
            if ch != "<":
                self._emit_text(ch)
                i += 1
                continue
            close = data.find(">", i + 1)
            if close < 0:
                # Incomplete tag — hold as carry unless finalizing.
                if not finalize:
                    self._carry = data[i:]
                    return
                self._emit_text(data[i:])
                return
            inner = data[i + 1 : close]
            classified = _classify_tag(inner)
            if classified is None:
                self._emit_text(data[i : close + 1])
                i = close + 1
                continue
            name, is_closing, noncanonical = classified
            if noncanonical:
                self._errors.add("noncanonical_reserved_tag")
            self._handle_tag(name, is_closing)
            i = close + 1

    def _handle_tag(self, name: str, is_closing: bool) -> None:
        if self._discard_until is not None:
            expect = self._discard_until
            if is_closing and name == expect:
                self._discard_until = None
                if name == "followups" and self._stack and self._stack[-1] == "think":
                    # Nested followups closed while think still open.
                    pass
                elif name == "think" and self._stack and self._stack[-1] == "followups":
                    pass
                return
            if not is_closing:
                # Nested opens while discarding stay discarded.
                return
            # Wrong close while discarding: may be mismatched think close.
            if (
                expect == "followups"
                and is_closing
                and name == "think"
                and self._stack
                and self._stack[-1] == "think"
            ):
                self._errors.add("mismatched_close")
                self._close_think_for_stream()
                self._stack.pop()
                # Keep discarding until the nested followups close.
                return
            if (
                expect == "think"
                and is_closing
                and name == "followups"
                and self._stack
                and self._stack[-1] == "followups"
            ):
                self._errors.add("mismatched_close")
                self._stack.pop()
                self._followups_invalid = True
                self._discard_until = None
                return
            return

        if not is_closing:
            self._handle_open(name)
            return
        self._handle_close(name)

    def _handle_open(self, name: str) -> None:
        if self._stack:
            top = self._stack[-1]
            if top != name:
                self._errors.add("nested_reserved_tag")
                if top == "think" and name == "followups":
                    self._followups_invalid = True
                    self._discard_until = "followups"
                    return
                if top == "followups" and name == "think":
                    self._followups_invalid = True
                    self._discard_until = "think"
                    return
            else:
                # Same tag nested — treat as nested error and discard until close.
                self._errors.add("nested_reserved_tag")
                self._discard_until = name
                if name == "followups":
                    self._followups_invalid = True
                return

        if name == "think":
            self._stack.append("think")
            self._safe_stream += "<think>"
            self._think_open_emitted = True
            return

        # followups
        self._stack.append("followups")
        self._followup_buf = ""
        return

    def _handle_close(self, name: str) -> None:
        if not self._stack:
            self._errors.add("stray_close")
            return
        top = self._stack[-1]
        if top != name:
            self._errors.add("mismatched_close")
            # Remove the mismatched unclosed tag so parsing can converge.
            if name in self._stack:
                while self._stack and self._stack[-1] != name:
                    removed = self._stack.pop()
                    if removed == "think":
                        self._close_think_for_stream()
                    elif removed == "followups":
                        self._followups_invalid = True
                if self._stack and self._stack[-1] == name:
                    self._stack.pop()
                    if name == "think":
                        self._close_think_for_stream()
                    elif name == "followups":
                        self._finish_followups_block()
            else:
                # Close for a tag not on stack: stray after mismatch bookkeeping.
                self._errors.add("stray_close")
            return

        self._stack.pop()
        if name == "think":
            self._close_think_for_stream()
            return
        self._finish_followups_block()

    def _close_think_for_stream(self) -> None:
        if self._think_open_emitted:
            self._safe_stream += "</think>"
            self._think_open_emitted = False

    def _finish_followups_block(self) -> None:
        self._followups_blocks += 1
        if self._followups_blocks > 1:
            self._errors.add("multiple_followups")
            self._followups_invalid = True
        # Keep buffer for finalize; invalid/multiple cleared later.

    def _emit_text(self, text: str) -> None:
        if not text:
            return
        if self._discard_until is not None:
            return
        if not self._stack:
            self._body += text
            self._safe_stream += text
            return
        top = self._stack[-1]
        if top == "think":
            self._reasoning += text
            self._safe_stream += text
            return
        if top == "followups":
            self._followup_buf += text
            return

    def _finalize_errors_and_suggestions(self) -> None:
        if self._discard_until == "followups":
            self._errors.add("unclosed_followups")
        elif self._discard_until == "think":
            self._errors.add("unclosed_think")
        for name in list(self._stack):
            if name == "think":
                self._errors.add("unclosed_think")
            elif name == "followups":
                self._errors.add("unclosed_followups")
        self._stack.clear()
        self._close_think_for_stream()

    def _to_parsed(self) -> ParsedAssistantOutput:
        body = self._body
        reasoning = self._reasoning
        errors = self._errors

        suggestions: tuple[str, ...] = ()
        had_followups_attempt = self._followups_blocks > 0 or bool(
            self._followup_buf.strip()
        )

        if self._followups_blocks > 1 or self._followups_invalid:
            suggestions = ()
        elif self._followups_blocks == 1 and "unclosed_followups" not in errors.as_tuple():
            raw_lines = self._followup_buf
            if not body.strip():
                errors.add("missing_visible_body")
                suggestions = ()
            else:
                suggestions = sanitize_suggested_questions(raw_lines, body)
                if not suggestions:
                    # Distinguish count vs line issues for protocol codes.
                    nonempty = [
                        strip_model_control_artifacts(x.strip())
                        for x in raw_lines.splitlines()
                        if strip_model_control_artifacts(x.strip())
                    ]
                    if len(nonempty) != 3:
                        errors.add("invalid_followup_count")
                    else:
                        errors.add("invalid_followup_line")
        elif had_followups_attempt and not body.strip() and self._followups_blocks == 0:
            # Unclosed followups already recorded; no missing_visible_body.
            pass

        if errors.as_tuple():
            # Malformed turns never expose suggestions.
            suggestions = ()

        return ParsedAssistantOutput(
            reasoning=reasoning,
            visible_body=body,
            suggested_questions=suggestions,
            protocol_errors=errors.as_tuple(),
        )


def parse_assistant_output(raw: str) -> ParsedAssistantOutput:
    """Parse a full assistant raw string into canonical protocol fields."""
    parser = AssistantOutputStreamParser()
    parser.feed(str(raw or ""))
    return parser.finalize()

File: desktop/src/utils/assistant-output.ts
/**
 * Canonical assistant output parser (think / followups protocol).
 * Mirrors agenticx/runtime/assistant_output.py — keep both in sync via the
 * shared golden fixture under tests/fixtures/assistant_output_protocol_cases.json.
 *
 * Author: Damon Li
 */

const MINIMAX_ARTIFACT_RE =
  /[\]\[<>~!|]+\s*\/?\s*minimax[a-z_:]*\s*[\]\[<>~!|]*/gi;
const ZERO_WIDTH_RE = /[\u200b\u200c\u200d\u2060\ufeff\u00ad]/g;
const RESERVED_TAG_TOKEN_RE = /<\/?\s*think\b|<\/?\s*followups\b/i;
const CONTROL_CHAR_RE = /[\x00-\x08\x0b\x0c\x0e-\x1f\x7f]/;
const TAG_NAME_RE = /^(?<closing>\/?)\s*(?<name>think|followups)\b(?<rest>.*)$/is;

export type ParsedAssistantOutputForUi = {
  reasoning: string;
  visibleBody: string;
  suggestedQuestions: string[];
  protocolErrors: string[];
  malformed: boolean;
};

export function stripModelControlArtifacts(text: string): string {
  if (!text || !text.toLowerCase().includes("minimax")) return text;
  return text.replace(MINIMAX_ARTIFACT_RE, "").trim();
}

function isMarkdownArtifactLine(line: string): boolean {
  const s = line.trim();
  if (!s) return true;
  if (s.startsWith("```") || s.endsWith("```")) return true;
  if (s.startsWith("#")) return true;
  if (/^\|.*\|$/.test(s)) return true;
  if (/^\|?\s*:?-{3,}:?\s*(\|\s*:?-{3,}:?\s*)+\|?$/.test(s)) return true;
  return false;
}

export function sanitizeSuggestedQuestions(
  raw: unknown,
  visibleBody: string,
): string[] {
  if (!String(visibleBody ?? "").trim()) return [];
  let candidates: string[];
  if (typeof raw === "string") {
    candidates = raw.split(/\r?\n/);
  } else if (Array.isArray(raw)) {
    candidates = raw.map((item) => String(item));
  } else {
    return [];
  }
  const lines: string[] = [];
  for (const item of candidates) {
    const trimmed = stripModelControlArtifacts(String(item).trim());
    if (trimmed) lines.push(trimmed);
  }
  if (lines.length !== 3) return [];
  for (const line of lines) {
    if (line.length < 1 || line.length > 160) return [];
    if (RESERVED_TAG_TOKEN_RE.test(line)) return [];
    if (isMarkdownArtifactLine(line)) return [];
  }
  return lines;
}

function normalizeTagInner(inner: string): string {
  return inner.replace(ZERO_WIDTH_RE, "").trim();
}

function classifyTag(
  inner: string,
): { name: string; isClosing: boolean; noncanonical: boolean } | null {
  const normalized = normalizeTagInner(inner);
  if (!normalized) return null;
  const match = TAG_NAME_RE.exec(normalized);
  if (!match || !match.groups) return null;
  const name = match.groups.name.toLowerCase();
  const isClosing = Boolean(match.groups.closing);
  const rest = match.groups.rest ?? "";
  const compactCanonical = `${isClosing ? "/" : ""}${name}`;
  const originalCompact = inner.replace(ZERO_WIDTH_RE, "").replace(/\s+/g, "").toLowerCase();
  let noncanonical = originalCompact !== compactCanonical || Boolean(rest.trim());
  if (inner !== inner.trim() || inner.includes("\n") || inner.includes("\r")) {
    noncanonical = true;
  }
  if (ZERO_WIDTH_RE.test(inner)) noncanonical = true;
  return { name, isClosing, noncanonical };
}

class ProtocolErrorSet {
  private items: string[] = [];
  private seen = new Set<string>();

  add(code: string): void {
    if (this.seen.has(code)) return;
    this.seen.add(code);
    this.items.push(code);
  }

  asArray(): string[] {
    return [...this.items];
  }
}

class StreamParser {
  private carry = "";
  private stack: string[] = [];
  private body = "";
  private reasoning = "";
  private followupBuf = "";
  private followupsBlocks = 0;
  private followupsInvalid = false;
  private discardUntil: string | null = null;
  private errors = new ProtocolErrorSet();
  private safeStream = "";
  private thinkOpenEmitted = false;

  feed(chunk: string): string {
    if (!chunk) return "";
    const before = this.safeStream.length;
    this.consume(this.carry + chunk, false);
    return this.safeStream.slice(before);
  }

  finalize(): ParsedAssistantOutputForUi {
    if (this.carry) {
      this.emitText(this.carry);
      this.carry = "";
    }
    this.finalizeErrors();
    return this.toParsed();
  }

  private consume(data: string, finalize: boolean): void {
    let i = 0;
    this.carry = "";
    while (i < data.length) {
      if (data[i] !== "<") {
        this.emitText(data[i]!);
        i += 1;
        continue;
      }
      const close = data.indexOf(">", i + 1);
      if (close < 0) {
        if (!finalize) {
          this.carry = data.slice(i);
          return;
        }
        this.emitText(data.slice(i));
        return;
      }
      const inner = data.slice(i + 1, close);
      const classified = classifyTag(inner);
      if (!classified) {
        this.emitText(data.slice(i, close + 1));
        i = close + 1;
        continue;
      }
      if (classified.noncanonical) this.errors.add("noncanonical_reserved_tag");
      this.handleTag(classified.name, classified.isClosing);
      i = close + 1;
    }
  }

  private handleTag(name: string, isClosing: boolean): void {
    if (this.discardUntil !== null) {
      const expect = this.discardUntil;
      if (isClosing && name === expect) {
        this.discardUntil = null;
        return;
      }
      if (!isClosing) return;
      if (
        expect === "followups" &&
        isClosing &&
        name === "think" &&
        this.stack.length > 0 &&
        this.stack[this.stack.length - 1] === "think"
      ) {
        this.errors.add("mismatched_close");
        this.closeThinkForStream();
        this.stack.pop();
        return;
      }
      if (
        expect === "think" &&
        isClosing &&
        name === "followups" &&
        this.stack.length > 0 &&
        this.stack[this.stack.length - 1] === "followups"
      ) {
        this.errors.add("mismatched_close");
        this.stack.pop();
        this.followupsInvalid = true;
        this.discardUntil = null;
        return;
      }
      return;
    }
    if (!isClosing) {
      this.handleOpen(name);
      return;
    }
    this.handleClose(name);
  }

  private handleOpen(name: string): void {
    if (this.stack.length > 0) {
      const top = this.stack[this.stack.length - 1]!;
      if (top !== name) {
        this.errors.add("nested_reserved_tag");
        if (top === "think" && name === "followups") {
          this.followupsInvalid = true;
          this.discardUntil = "followups";
          return;
        }
        if (top === "followups" && name === "think") {
          this.followupsInvalid = true;
          this.discardUntil = "think";
          return;
        }
      } else {
        this.errors.add("nested_reserved_tag");
        this.discardUntil = name;
        if (name === "followups") this.followupsInvalid = true;
        return;
      }
    }
    if (name === "think") {
      this.stack.push("think");
      this.safeStream += "<think>";
      this.thinkOpenEmitted = true;
      return;
    }
    this.stack.push("followups");
    this.followupBuf = "";
  }

  private handleClose(name: string): void {
    if (this.stack.length === 0) {
      this.errors.add("stray_close");
      return;
    }
    const top = this.stack[this.stack.length - 1]!;
    if (top !== name) {
      this.errors.add("mismatched_close");
      if (this.stack.includes(name)) {
        while (this.stack.length > 0 && this.stack[this.stack.length - 1] !== name) {
          const removed = this.stack.pop();
          if (removed === "think") this.closeThinkForStream();
          else if (removed === "followups") this.followupsInvalid = true;
        }
        if (this.stack.length > 0 && this.stack[this.stack.length - 1] === name) {
          this.stack.pop();
          if (name === "think") this.closeThinkForStream();
          else if (name === "followups") this.finishFollowupsBlock();
        }
      } else {
        this.errors.add("stray_close");
      }
      return;
    }
    this.stack.pop();
    if (name === "think") {
      this.closeThinkForStream();
      return;
    }
    this.finishFollowupsBlock();
  }

  private closeThinkForStream(): void {
    if (this.thinkOpenEmitted) {
      this.safeStream += "</think>";
      this.thinkOpenEmitted = false;
    }
  }

  private finishFollowupsBlock(): void {
    this.followupsBlocks += 1;
    if (this.followupsBlocks > 1) {
      this.errors.add("multiple_followups");
      this.followupsInvalid = true;
    }
  }

  private emitText(text: string): void {
    if (!text) return;
    if (this.discardUntil !== null) return;
    if (this.stack.length === 0) {
      this.body += text;
      this.safeStream += text;
      return;
    }
    const top = this.stack[this.stack.length - 1]!;
    if (top === "think") {
      this.reasoning += text;
      this.safeStream += text;
      return;
    }
    if (top === "followups") {
      this.followupBuf += text;
    }
  }

  private finalizeErrors(): void {
    if (this.discardUntil === "followups") this.errors.add("unclosed_followups");
    else if (this.discardUntil === "think") this.errors.add("unclosed_think");
    for (const name of [...this.stack]) {
      if (name === "think") this.errors.add("unclosed_think");
      else if (name === "followups") this.errors.add("unclosed_followups");
    }
    this.stack = [];
    this.closeThinkForStream();
  }

  private toParsed(): ParsedAssistantOutputForUi {
    const errors = this.errors;
    let suggestions: string[] = [];
    if (this.followupsBlocks > 1 || this.followupsInvalid) {
      suggestions = [];
    } else if (
      this.followupsBlocks === 1 &&
      !errors.asArray().includes("unclosed_followups")
    ) {
      if (!this.body.trim()) {
        errors.add("missing_visible_body");
        suggestions = [];
      } else {
        suggestions = sanitizeSuggestedQuestions(this.followupBuf, this.body);
        if (suggestions.length === 0) {
          const nonempty = this.followupBuf
            .split(/\r?\n/)
            .map((x) => stripModelControlArtifacts(x.trim()))
            .filter(Boolean);
          if (nonempty.length !== 3) errors.add("invalid_followup_count");
          else errors.add("invalid_followup_line");
        }
      }
    }
    const protocolErrors = errors.asArray();
    if (protocolErrors.length > 0) suggestions = [];
    return {
      reasoning: this.reasoning,
      visibleBody: this.body,
      suggestedQuestions: suggestions,
      protocolErrors,
      malformed: protocolErrors.length > 0,
    };
  }
}

export function parseAssistantOutputForUi(raw: string): ParsedAssistantOutputForUi {
  const parser = new StreamParser();
  parser.feed(String(raw ?? ""));
  return parser.finalize();
}

export function assistantVisibleBodyForUi(content: string): string {
  return parseAssistantOutputForUi(content).visibleBody;
}

export function normalizeFinalAssistantPayload(
  rawText: unknown,
  rawQuestions: unknown,
  rawTurnTerminal: unknown,
  rawTerminalReason: unknown,
): {
  text: string;
  suggestedQuestions: string[];
  incomplete: boolean;
  turnTerminal: boolean;
  terminalReason?: string;
} {
  const parsed = parseAssistantOutputForUi(String(rawText ?? ""));
  const text = parsed.visibleBody.trim();
  const terminalReason = String(rawTerminalReason ?? "").trim() || undefined;
  return {
    text,
    suggestedQuestions: parsed.malformed
      ? []
      : sanitizeSuggestedQuestions(rawQuestions, text),
    incomplete: !text,
    turnTerminal:
      rawTurnTerminal === true
        ? true
        : rawTurnTerminal === false
          ? false
          : text.length > 0,
    terminalReason,
  };
}
